IdentityCore/src/validation/MSIDAdfsAuthorityResolver.m

```diff
@@ -0,0 +1,182 @@
+         [webFingerRequest sendWithBlock:^(id response, NSError *error)
+          {
+              if (error)
+              {
+                  if (completionBlock) completionBlock(nil, NO, error);
+                  return;
+              }
+              
+              if ([self isRealmTrustedFromWebFingerPayload:response authority:authority])
+              {
+                  __auto_type openIdConfigurationEndpoint = [self openIdConfigurationEndpointForAuthority:authority];
+                  
+                  __auto_type cacheRecord = [MSIDAuthorityCacheRecord new];
+                  cacheRecord.validated = YES;
+                  cacheRecord.openIdConfigurationEndpoint = openIdConfigurationEndpoint;
+                  [s_cache setObject:cacheRecord forKey:authority.absoluteString.lowercaseString];
+                  
+                  if (completionBlock) completionBlock(openIdConfigurationEndpoint, YES, nil);
+              }
+              else
+              {
+                  error = MSIDCreateError(MSIDErrorDomain, MSIDErrorDeveloperAuthorityValidation, @"WebFinger request was invalid or failed", nil, nil, nil, context.correlationId, nil);
+                  if (completionBlock) completionBlock(nil, NO, error);
+              }
+          }];
+     }];
+}
+
+#pragma mark - Private
+
+- (void)sendDrsDiscoveryWithDomain:(NSString *)domain
+                           context:(id<MSIDRequestContext>)context
+                   completionBlock:(MSIDHttpRequestDidCompleteBlock)completionBlock
+{
+    __auto_type drsPremRequest = [[MSIDDRSDiscoveryRequest alloc] initWithDomain:domain adfsType:MSIDADFSTypeOnPrems];
+    drsPremRequest.context = context;
+    [drsPremRequest sendWithBlock:^(id response, NSError *error)
+     {
+         if (response)
+         {
+             if (completionBlock) completionBlock(response, error);
+             return;
+         }
+         
+         __auto_type drsCloudRequest = [[MSIDDRSDiscoveryRequest alloc] initWithDomain:domain adfsType:MSIDADFSTypeCloud];
+         drsCloudRequest.context = context;
+         [drsCloudRequest sendWithBlock:^(id response, NSError *error)
+          {
+              if (response)
+              {
+                  if (completionBlock) completionBlock(response, error);
+                  return;
+              }
+          }];
+     }];
+}
+
+- (BOOL)isRealmTrustedFromWebFingerPayload:(id)json
+                                 authority:(NSURL *)authority
+{
+    NSArray *links = [json objectForKey:@"links"];
+    for (id link in links)
+    {
+        NSString *rel = [link objectForKey:@"rel"];
+        NSString *target = [link objectForKey:@"href"];
+        
+        NSURL *targetURL = [NSURL URLWithString:target];
+        
+        if ([rel caseInsensitiveCompare:s_kTrustedRelation] == NSOrderedSame &&
+            [targetURL msidIsEquivalentAuthority:authority])
```

Comment:
this should be msidIsEquivalentAuthorityHost, because the path will always mismatch

Reply:
done